scripts/generate-plugins.go

```diff
@@ -0,0 +1,273 @@
+//
+// Which is long, but deterministic and unique.
+func deriveImport(typeName, derivedName string) string {
+	return strings.Replace(derivedName, "-", "", -1) + strings.ToLower(typeName)
+}
+
+// discoverTypesInPath searches for types of typeID in path and returns a list
+// of plugins it finds.
+func discoverTypesInPath(path, typeID, typeName string) ([]plugin, error) {
+	pluginTypes := []plugin{}
+
+	dirs, err := listDirectories(path)
+	if err != nil {
+		return pluginTypes, err
+	}
+
+	for _, dir := range dirs {
+		fset := token.NewFileSet()
+		goPackages, err := parser.ParseDir(fset, dir, nil, parser.AllErrors)
+		if err != nil {
+			return pluginTypes, fmt.Errorf("Failed parsing directory %s: %s", dir, err)
+		}
+
+		for _, goPackage := range goPackages {
+			ast.PackageExports(goPackage)
+			ast.Inspect(goPackage, func(n ast.Node) bool {
+				switch x := n.(type) {
+				case *ast.FuncDecl:
+					// If we get a function then we will check the function name
+					// against typeName and the function return type (Results)
+					// against typeID.
+					//
+					// There may be more than one return type but in the target
+					// case there should only be one. Also the return type is a
+					// ast.SelectorExpr which means we have multiple nodes.
+					// We'll read all of them as ast.Ident (identifier), join
+					// them via . to get a string like terraform.ResourceProvider
+					// and see if it matches our expected typeID
+					//
+					// This is somewhat verbose but prevents us from identifying
+					// the wrong types if the function name is amiguous or if
+					// there are other subfolders added later.
+					if x.Name.Name == typeName && len(x.Type.Results.List) == 1 {
+						node := x.Type.Results.List[0].Type
+						typeIdentifiers := []string{}
+						ast.Inspect(node, func(m ast.Node) bool {
+							switch y := m.(type) {
+							case *ast.Ident:
+								typeIdentifiers = append(typeIdentifiers, y.Name)
+							}
+							// We need all of the identifiers to join so we
+							// can't break early here.
+							return true
+						})
+						if strings.Join(typeIdentifiers, ".") == typeID {
+							derivedName := deriveName(path, dir)
+							pluginTypes = append(pluginTypes, plugin{
+								Package:    goPackage.Name,
+								PluginName: derivedName,
+								ImportName: deriveImport(x.Name.Name, derivedName),
+								TypeName:   x.Name.Name,
+								Path:       dir,
+							})
+						}
+					}
+				case *ast.TypeSpec:
+					if x.Name.Name == typeID {
+						derivedName := deriveName(path, dir)
+						pluginTypes = append(pluginTypes, plugin{
+							Package:    goPackage.Name,
+							PluginName: derivedName,
+							ImportName: deriveImport(x.Name.Name, derivedName),
+							TypeName:   x.Name.Name,
+							Path:       dir,
+						})
+						// The AST stops parsing when we return false. Once we
+						// find the symbol we want we can stop parsing.
+						return false
+					}
+				}
+				return true
+			})
+		}
+	}
```

Comment:
We might want to add a unit test or three around this behavior at some point, but since we'll be exercising this code quite a bit I'm cool going with it for now.